File: src/simulation_parameters/four_parameters.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class FourParameters:

    fundamental_frequency: str
    output_variables: tuple[str, ...]

    @classmethod
    def from_xyce_statement(cls, four_statement: str) -> "FourParameters" | None:
        # tokenize the directive (simple split for now)
        tokens = four_statement.split()
        # reject non-four statements
        if len(tokens) < 3 or tokens[0].upper() != ".FOUR":
            return None
        # fundamental frequency
        fundamental_frequency = tokens[1]
        # output variables
        output_variables = tuple(tokens[2:])
        # return model
        return cls(fundamental_frequency=fundamental_frequency, output_variables=output_variables)

    def to_xyce_statement(self) -> str:
        # init tokens
        tokens = [".FOUR", self.fundamental_frequency]
        # add output variables
        tokens.extend(self.output_variables)
        # return joined statement
        return " ".join(tokens)

File: src/simulation_parameters/four_parameters.cpp
#include <cctype>
#include <string>
#include <vector>

#include "../util.h"
#include "four_parameters.h"

FourParameters::FourParameters(std::string fundamental_frequency, std::vector<std::string> output_variables) :
    fundamental_frequency(std::move(fundamental_frequency)), output_variables(std::move(output_variables)) {}

std::optional<FourParameters> FourParameters::from_xyce_statement(const std::string& four_statement) {
    // tokenize the directive respecting brace-enclosed expressions
    std::vector<std::string> tokens;
    std::string current;
    int brace_depth = 0;
    // iterate characters
    for (const char ch : four_statement) {
        // check opening brace
        if (ch == '{') {
            current += ch;
            ++brace_depth;
            continue;
        }
        // check closing brace
        if (ch == '}') {
            current += ch;
            --brace_depth;
            continue;
        }
        // check whitespace splitter when not inside braces
        if (std::isspace(static_cast<unsigned char>(ch)) && brace_depth == 0) {
            // flush current token when non-empty
            if (!current.empty()) {
                tokens.push_back(current);
                current.clear();
            }
            continue;
        }
        // append char
        current += ch;
    }
    // flush trailing token
    if (!current.empty()) {
        tokens.push_back(current);
    }
    // reject non-four statements
    if (tokens.size() < 3 || to_upper(tokens[0]) != ".FOUR") {
        return std::nullopt;
    }
    // fundamental frequency is the second token
    const std::string fundamental_frequency = tokens[1];
    // output variables are all remaining tokens
    std::vector<std::string> output_variables(tokens.begin() + 2, tokens.end());
    // return model
    return FourParameters(fundamental_frequency, std::move(output_variables));
}

std::string FourParameters::to_xyce_statement() const {
    // init tokens
    std::vector<std::string> tokens = {".FOUR", fundamental_frequency};
    // add output variables
    for (const auto& var : output_variables) {
        tokens.push_back(var);
    }
    // build joined statement
    std::string result;
    for (size_t i = 0; i < tokens.size(); ++i) {
        // add separator
        if (i > 0) {
            result += ' ';
        }
        // add token
        result += tokens[i];
    }
    // return joined statement
    return result;
}

bool FourParameters::operator==(const FourParameters& other) const {
    // compare all fields for equality
    return fundamental_frequency == other.fundamental_frequency && output_variables == other.output_variables;
}
